pkg/ccl/crosscluster/logical/lww_row_processor_test.go

```diff
@@ -301,3 +306,170 @@ func BenchmarkLWWInsertBatch(b *testing.B) {
+
+		if useKVProc {
+			rp, err = newKVRowProcessor(ctx,
+				&execinfra.ServerConfig{
+					DB:           s.InternalDB().(descs.DB),
+					LeaseManager: s.LeaseManager(),
+				}, &eval.Context{
+					Codec:    s.Codec(),
+					Settings: s.ClusterSettings(),
+				}, map[descpb.ID]sqlProcessorTableConfig{
+					dstDesc.GetID(): {
+						srcDesc: srcDesc,
+					},
+				},
+				rp.(*sqlRowProcessor))
+			require.NoError(t, err)
+		}
+		return tableNameDst, rp, func(originTimestamp hlc.Timestamp, datums ...interface{}) roachpb.KeyValue {
+			kv := replicationtestutils.EncodeKV(t, s.Codec(), srcDesc, datums...)
+			kv.Value.Timestamp = originTimestamp
+			return kv
+		}
+	}
+
+	insertRow := func(rp RowProcessor, keyValue roachpb.KeyValue, prevValue roachpb.Value) error {
+		return s.InternalDB().(isql.DB).Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			_, err := rp.ProcessRow(ctx, txn, keyValue, prevValue)
+			return err
+		})
+	}
+
+	timeNow := hlc.Timestamp{WallTime: timeutil.Now().UnixNano()}
+	timeNowPlusOne := hlc.Timestamp{WallTime: timeutil.Now().Add(time.Microsecond).UnixNano()}
+	timeOneDayForward := hlc.Timestamp{WallTime: timeutil.Now().Add(time.Hour * 24).UnixNano()}
+	timeOneDayBackward := hlc.Timestamp{WallTime: timeutil.Now().Add(time.Hour * -24).UnixNano()}
+	row1 := []interface{}{1, "row1"}
+	row2 := []interface{}{1, "row2"}
+	row3 := []interface{}{1, "row3"}
+
+	// Run with both the SQL row processor and the KV row processor. Note that currently,
+	// KV row proc write conflicts result in a fallback to the SQL row processor.
+	testutils.RunTrueAndFalse(t, "useKVProc", func(t *testing.T, useKVProc bool) {
+		// All of those combinations are tested with both optimistic inserts and standard inserts.
+		testutils.RunTrueAndFalse(t, "optimistic_insert", func(t *testing.T, optimisticInsert bool) {
+
+			// Write to table on both the src and dest cluster and see how conflicts are handled
+			// The local write will not set crdb_replication_origin_timestamp
+			t.Run("cross-cluster-insert", func(t *testing.T) {
+				runner.Exec(t, fmt.Sprintf("SET CLUSTER SETTING logical_replication.consumer.try_optimistic_insert.enabled=%t", optimisticInsert))
+				tableNameDst, rp, encoder := setup(t, useKVProc)
+
+				runner.Exec(t, fmt.Sprintf("INSERT INTO %s VALUES ($1, $2)", tableNameDst), row1...)
+
+				keyValue2 := encoder(timeOneDayBackward, row2...)
+				require.NoError(t, insertRow(rp, keyValue2, roachpb.Value{}))
+
+				expectedRows := [][]string{
+					{"1", "row1"},
+				}
+				runner.CheckQueryResults(t, fmt.Sprintf("SELECT * from %s", tableNameDst), expectedRows)
+
+				keyValue3 := encoder(timeOneDayForward, row2...)
+				require.NoError(t, insertRow(rp, keyValue3, keyValue2.Value))
+
+				expectedRows = [][]string{
+					{"1", "row2"},
+				}
+				runner.CheckQueryResults(t, fmt.Sprintf("SELECT * from %s", tableNameDst), expectedRows)
+			})
+
+			// Receive multiple updates remotely and handle conflicts between them
```

Comment:
docstring could be sharper: when a remote update conflicts with another remote update, LWW uses their OriginTimestamps.